### PR TITLE
refactor kernels.py into submodule

### PR DESCRIPTION
**Pull request by db091756**
### PR Type

- Refactoring (no functional changes)

### Description

Creates `kernels` subpackage.

`kernels.util.py` holds utility functions relating to kernels.
`kernels.base.py` holds the base kernel class `ScalarValuedKernel`, as well as the base composite classes `UniCompositeKernel`,  `DuoCompositeKernel` and the derived duo-composite kernels `AdditiveKernel` and `ProductKernel`
`kernels.scalar_valued.py` holds all currently implemented scalar valued kernels.

### How Has This Been Tested?

All existing tests pass.

### Does this PR introduce a breaking change?

(Write your answer here.)

### Screenshots

(Write your answer here.)

### Checklist before requesting a review
- [x] I have made sure that my PR is not a duplicate.
- [x] My code follows the style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have performed a self-review of my code.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

File: coreax/kernels/base.py [hunk 1]
```diff
@@ -0,0 +1,503 @@
...
+
+
+class ScalarValuedKernel(eqx.Module):
+    """Abstract base class for scalar-valued kernels."""
+
+    def __add__(self, addition: ScalarValuedKernel) -> AdditiveKernel:
+        """Overload `+` operator."""
+        if not isinstance(addition, ScalarValuedKernel):
+            raise ValueError("'addition' must be an instance of a 'ScalarValuedKernel'")
+        return AdditiveKernel(self, addition)
```
Comment (rg936672) [inline coreax/kernels/base.py:78]: We've lost the ability to directly add a scalar to a kernel. (Also, if that's added back in, we'll need to reimplement `__radd__` to allow `kernel + scalar` as well as `scalar + kernel`).

File: coreax/kernels/base.py [hunk 2]
```diff
@@ -0,0 +1,503 @@
...
+
+    def __mul__(self, product: ScalarValuedKernel) -> ProductKernel:
+        """Overload `*` operator."""
+        if not isinstance(product, ScalarValuedKernel):
+            raise ValueError("'product' must be an instance of a 'ScalarValuedKernel'")
+        return ProductKernel(self, product)
```
Comment (rg936672) [inline coreax/kernels/base.py:84]: We've also lost the ability to directly multiply a kernel by a scalar. (And we'll need to reimplement `__rmul__` to make it work both ways).

File: coreax/kernels/base.py [hunk 3]
```diff
@@ -0,0 +1,503 @@
...
+
+        # Ensure the first and second kernels are symmetric for even powers to make use
+        # of reduced computation capabilities in ProductKernel. For example,
+        # :meth:`~divergence_x_grad_y_elementwise` can be computed more efficiently if
+        # the first and second kernels are recognised as the same.
+        for i in range(min_power, power):
+            if i % 2 == 0:
+                first_kernel = ProductKernel(first_kernel, self)
+            else:
+                second_kernel = ProductKernel(second_kernel, self)
```
Comment (rg936672) [inline coreax/kernels/base.py:104]: Would it be more efficient to instead implement a `PowerKernel` subclassing `UniCompositeKernel` that computes this directly, rather than relying on nested `ProductKernel` instances? That would also allow non-integer powers. That's probably out-of-scope for this PR, though, so may want to be its own issue.

File: coreax/kernels/base.py [hunk 4]
```diff
@@ -0,0 +1,503 @@
...
+            block_sum, jnp.asarray(0.0), block_x, unroll=outer_unroll
+        )
+        if axis is None:
+            return accumulated_sum
+        num_rows_padded = block_x.data.shape[0] * block_x.data.shape[1]
+        column_sum_padded = column_sum_blocks.reshape(num_rows_padded, -1).sum(axis=1)
+        return column_sum_padded[:unpadded_len_x]
+
+
+class UniCompositeKernel(ScalarValuedKernel):
```
Comment (rg936672) [inline coreax/kernels/base.py:350]: Would it make more sense for this (and `DuoCompositeKernel`, `AdditiveKernel`, `ProductKernel`) to be in `coreax.kernels.scalar_valued` instead? Presumably when vector/function-valued kernels are introduced, we'll want to also have an `AdditiveKernel` (etc.) equivalent for them as well.

File: CHANGELOG.md [hunk 5]
```diff
@@ -20,6 +20,16 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   - `coreax.inverses.RegularisedInverseApproximator` replaced by `coreax.least_squares.RegularisedLeastSquaresSolver`
   - `coreax.inverses.LeastSquaresApproximator` replaced by `coreax.least_squares.MinimalEuclideanNormSolver`
   - `coreax.inverses.RandomisedEigendecompositionApproximator` replaced by `coreax.least_squares.RandomisedEigendecompositionSolver`
+- Refactoring of `coreax.kernel.py` into `coreax.kernels` sub-package:
+  - `kernels.util.py` holds utility functions relating to kernels e.g. `median_heuristic`.
```
Comment (rg936672) [inline CHANGELOG.md:24]: Mention that `block_data_convert` is now public.

File: coreax/kernels/base.py (hunk 1 above)
Comment (rg936672) [reply coreax/kernels/base.py:78]: Note that this (and the change to multiplication below) is a breaking change. Thus, it should have a deprecation period. (Or just not be changed - this feels like an unnecessary feature removal!)

File: coreax/kernels/base.py (hunk 4 above)
Comment (rg936672) [reply coreax/kernels/base.py:350]: (If moved, remember to change the changelog comment as well!)

**Review by rg936672 — CHANGES_REQUESTED** [review]
Generally looks good, but several changes here are breaking changes, which will need to have deprecation notices.

File: coreax/kernels/base.py [hunk 6]
```diff
@@ -0,0 +1,503 @@
...
+    def __mul__(self, product: ScalarValuedKernel) -> ProductKernel:
+        """Overload `*` operator."""
+        if not isinstance(product, ScalarValuedKernel):
+            raise ValueError("'product' must be an instance of a 'ScalarValuedKernel'")
+        return ProductKernel(self, product)
+
+    def __pow__(self, power: int) -> ProductKernel:
+        """Overload `**` operator."""
+        min_power = 2
+        power = int(power)
```
Comment (rg936672) [inline coreax/kernels/base.py:89]: Note that doing `int(power)` just rounds `power` to the nearest integer - it won't throw any kind of error. To see this, change the test for invalid inputs to `__pow__` to instead to `kernel ** 2.5`, and note that no error is thrown (as the float is silently rounded to an int).

File: coreax/kernels/base.py (hunk 2 above)
Comment (db091756) [reply coreax/kernels/base.py:84]: See above comment

File: coreax/kernels/base.py (hunk 3 above)
Comment (db091756) [reply coreax/kernels/base.py:104]: Yes that seems like a good idea, will try implement this

File: coreax/kernels/base.py (hunk 4 above)
Comment (db091756) [reply coreax/kernels/base.py:350]: If we move these classes then I have come into issues with circular imports as we have to import from the `scalar_valued.py` file which itself imports from `base.py` already. 

I couldn't come up with a way to have `AdditiveKernel` for example live outside the same class as `ScalarValuedKernel` for example

File: CHANGELOG.md (hunk 5 above)
Comment (db091756) [reply CHANGELOG.md:24]: Will make this a private function again

File: coreax/kernels/base.py (hunk 6 above)
Comment (db091756) [reply coreax/kernels/base.py:89]: I did realise this was happening but given that you cannot, in general, take fractional powers of kernels and [expect them to still be valid kernels](https://stats.stackexchange.com/questions/566822/in-machine-learning-for-a-kernel-function-k-is-sqrtk-also-a-valid-kernel-fun) I decided to round the float. I can see how this could be confusing and will make this error in the implementation of `PowerKernel`

Comment (rg936672) [reply coreax/kernels/base.py:89]: Ah, that's fair enough, I wasn't aware of that issue. Although just looking at the answer you linked, is the issue there that the exponent is less than 1, rather than that the exponent is fractional? I'm afraid I'm not familiar enough with kernels to answer that myself!

File: coreax/kernels/base.py (hunk 4 above)
Comment (rg936672) [reply coreax/kernels/base.py:350]: You _could_ do 

```python
# __init__.py
from coreax.kernels.base import ScalarValuedKernel
from coreax.kernels.scalar_valued import (AdditiveKernel, ProductKernel, ...)

# base.py
from coreax import kernels
def function(x: kernels.AdditiveKernel): ...

# scalar_valued.py
from coreax import kernels
def other(x: kernels.ScalarValuedKernel): ...
```

and this avoids the circular import issue with some import-logic wizardry.

Alternatively, and I think this is possibly a better solution, you could move `base.py` and `scalar_valued.py` to e.g. `scalar_valued/base.py` and `scalar_valued/kernels.py` respectively, so that in future the other-valued kernels could be in e.g. `function_valued/` and `vector_valued/`. Then when importing from elsewhere, rather than `from coreax.kernels import XYZ`, you instead have to specify and go `from coreax.kernels.scalar_valued import XYZ`. (Then we can separately have `coreax.kernels.scalar_valued.AdditiveKernel` and `coreax.kernels.vector_valued.AdditiveKernel`; if client code needs to use both in the same file they can always just `import coreax.kernels.scalar_valued.AdditiveKernel as ScalarAdditiveKernel`.

File: coreax/kernels/base.py (hunk 1 above)
Comment (rg936672) [reply coreax/kernels/base.py:78]: It's not necessarily a significant loss, but it _is_ a breaking change, so we shouldn't make it without a deprecation period. (In this particular case, we can just move `LinearKernel` back to `base.py`, and it then Just Works™.)

File: coreax/kernels/base.py (hunk 3 above)
Comment (db091756) [reply coreax/kernels/base.py:104]: Implemented this now

File: coreax/kernels/base.py (hunk 6 above)
Comment (db091756) [reply coreax/kernels/base.py]: The positive semi-definiteness of kernel functions is very related to the positive semi-definiteness of their resulting (kernel matrices)[https://en.wikipedia.org/wiki/Positive-definite_function]. So if we are considering fractional powers:

Given some kernel function $k:\mathcal{X}\times\mathcal{X}\to\mathbb{R}$ define

$$p(x,y) = k(x,y)^n$$

where

File: coreax/kernels/base.py (hunk 1 above)
Comment (db091756) [reply coreax/kernels/base.py:78]: Added a private `_Constant` class to get around this issue.

Comment (db091756) [reply coreax/kernels/base.py:78]: Similar to https://scikit-learn.org/stable/modules/generated/sklearn.gaussian_process.kernels.ConstantKernel.html

File: coreax/kernels/base.py (hunk 2 above)
Comment (db091756) [reply coreax/kernels/base.py:84]: Added back in

File: coreax/kernels/base.py (hunk 4 above)
Comment (db091756) [reply coreax/kernels/base.py:333]: I am having trouble doing `from coreax import kernels` within the `base.py` and `scalar_valued.py` files, it appears to not recognise the import. Perhaps we can revisit this when I finish the operator-valued kernel implementation to see if this is an issue.

File: CHANGELOG.md [hunk 7]
```diff
@@ -20,6 +20,16 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   - `coreax.inverses.RegularisedInverseApproximator` replaced by `coreax.least_squares.RegularisedLeastSquaresSolver`
   - `coreax.inverses.LeastSquaresApproximator` replaced by `coreax.least_squares.MinimalEuclideanNormSolver`
   - `coreax.inverses.RandomisedEigendecompositionApproximator` replaced by `coreax.least_squares.RandomisedEigendecompositionSolver`
+- Refactoring of `coreax.kernel.py` into `coreax.kernels` sub-package:
```
[the remark this answers is not in the document]
Comment (db091756) [reply CHANGELOG.md:26]: Done this now

File: coreax/kernels/base.py (hunk 4 above)
Comment (rg936672) [reply coreax/kernels/base.py:385]: Revisiting this with the vector/function-valued kernels seems sensible, so let's leave this until then.

File: coreax/kernels/base.py (hunk 6 above)
Comment (rg936672) [reply coreax/kernels/base.py]: Might be worth mentioning this in the `ScalarValuedKernel` or `ScalarValuedKernel.__pow__` docstring.

File: coreax/kernels/base.py [hunk 8]
```diff
@@ -347,6 +351,37 @@ def slice_sum(accumulated_sum: Array, y_block: Data) -> tuple[Array, Array]:
...
 
 
+class _Constant(ScalarValuedKernel):
+    r"""Define a helper class to add additional functionality to magic methods."""
+
+    constant: float = eqx.field(default=1, converter=float)
+
+    def __check_init__(self):
+        """Check that the constant is not negative."""
+        if self.constant < 0:
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/kernels/base.py:361]: Yes, added. Also added an additional parameter

File: coreax/kernels/scalar_valued.py [hunk 9]
```diff
@@ -30,28 +30,42 @@ class LinearKernel(ScalarValuedKernel):
...
+    Given :math:`\rho`=` ``output_scale``, :math:`a=` ``constant`` and
+    :math: `b =` ``offset``,  the linear kernel is
     defined as :math:`k: \mathbb{R}^d\times \mathbb{R}^d \to \mathbb{R}`,
-    :math:`k(x, y) = \rho x^Ty + c`.
+    :math:`k(x, y) = a + \rho (x-b)^T(y-b)`.
 
     :param output_scale: Kernel normalisation constant, :math:`\rho`
-    :param constant: Additive constant, :math:`c`
+    :param constant: Additive constant, :math:`a`, must be positive
+    :param offset: Offset term, :math:`b`
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/kernels/scalar_valued.py:40]: Will do

File: coreax/kernels/scalar_valued.py [hunk 10]
```diff
@@ -30,28 +30,42 @@ class LinearKernel(ScalarValuedKernel):
...
     """
 
     output_scale: float = eqx.field(default=1.0, converter=float)
     constant: float = eqx.field(default=0.0, converter=float)
+    offset: float = eqx.field(default=0.0, converter=float)
+
+    def __check_init__(self):
+        """Check that the additive constant is not negative."""
+        if self.constant < 0:
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/kernels/scalar_valued.py:49]: Yes, this would need adding to every kernel in the codebase though. I guess we have been assuming up to this point people choose sensible parameters for these things.

[the remark this answers is not in the document]
Comment (db091756) [reply coreax/kernels/scalar_valued.py:49]: I could add `__check_init__`'s to every kernel or remove the one I added to stay consistent?

**Review by rg936672 — APPROVED** [review]

File: coreax/kernels/scalar_valued.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/kernels/scalar_valued.py]: See #719

File: coreax/kernels/scalar_valued.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/kernels/scalar_valued.py]: See #718
